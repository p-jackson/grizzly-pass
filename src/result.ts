export interface Result<T, E> {
  map<T2>(f: (t: T) => T2): Result<T2, E>;
  mapErr<E2>(f: (e: E) => E2): Result<T, E2>;
  isOk(): boolean;
  isErr(): boolean;
  andThen<T2, E2>(f: (t: T) => Result<T2, E2>): Result<T2, E | E2>;
  orElse<T2, E2>(f: (e: E) => Result<T2, E2>): Result<T | T2, E2>;
  unsafeUnwrap(): T;
}

export class Ok<T, E> implements Result<T, E> {
  value: T;

  constructor(value: T) {
    this.value = value;
  }

  map<T2>(f: (t: T) => T2): Result<T2, E> {
    return new Ok<T2, E>(f(this.value));
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  mapErr<E2>(f: (e: E) => E2): Result<T, E2> {
    return new Ok<T, E2>(this.value);
  }

  isOk() {
    return true;
  }

  isErr() {
    return false;
  }

  andThen<T2, E2>(f: (t: T) => Result<T2, E2>): Result<T2, E | E2> {
    return f(this.value);
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  orElse<T2, E2>(f: (e: E) => Result<T2, E2>): Result<T | T2, E2> {
    return new Ok<T, E2>(this.value);
  }

  unsafeUnwrap(): T {
    return this.value;
  }
}

export class Err<T, E> implements Result<T, E> {
  error: E;

  constructor(error: E) {
    this.error = error;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  map<T2>(f: (t: T) => T2): Result<T2, E> {
    return new Err<T2, E>(this.error);
  }

  mapErr<E2>(f: (e: E) => E2): Result<T, E2> {
    return new Err<T, E2>(f(this.error));
  }

  isOk() {
    return false;
  }

  isErr() {
    return true;
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  andThen<T2, E2>(f: (t: T) => Result<T2, E2>): Result<T2, E | E2> {
    return new Err<T2, E>(this.error);
  }

  orElse<T2, E2>(f: (e: E) => Result<T2, E2>): Result<T | T2, E2> {
    return f(this.error);
  }

  unsafeUnwrap(): T {
    if (
      typeof this.error === "string" ||
      (this.error && typeof this.error.toString === "function")
    )
      throw new Error(`Unwrapped result with err: ${this.error}`);
    else throw new Error(`Unwrapped result with err`);
  }
}

export const ok = <T>(t: T) => new Ok<T, unknown>(t);

export const err = <E>(e: E) => new Err<unknown, E>(e);

export function fromList<T, E>(
  listOfResults: Result<T, E>[],
): Result<T[], E[]> {
  const errs = listOfResults.reduce((acc, r) => {
    if (_isErr(r)) return [...acc, r.error];
    else return acc;
  }, [] as E[]);
  if (errs.length) return new Err(errs);
  else return new Ok(listOfResults.map((r) => r.unsafeUnwrap()));
}

function _isErr<T, E>(result: Result<T, E>): result is Err<T, E> {
  return result.isErr();
}
