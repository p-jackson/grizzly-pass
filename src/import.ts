import moment from "moment";
import { uniqueId } from "./unique-id";
import { ok, err, fromList, type Result } from "./result";
import { statusIds } from "./types";
import type { Project, Label, Status } from "./types";

type ImportProject = {
  title: string;
  person: string;
  date: string;
  progress: number;
  health: Status;
  tags: string[];
};

type Output = {
  title?: string;
  projects: Project[];
  labels: Label[];
};

export function importFile(
  fileContents: string,
): Result<Output, string | string[]> {
  return parseFile(fileContents)
    .mapErr((e) => `File ${e}`)
    .andThen((data) =>
      fromList(data.map(validateProject)).mapErr((badProjects) =>
        badProjects.map((badProject, i) => `Project ${i + 1} ${badProject}`),
      ),
    )
    .map((projects) => [projects, generateLabels(projects)])
    .map(([projects, labels]: [ImportProject[], Label[]]) => ({
      title: undefined,
      projects: projects.map((project) => {
        const { title, person, health, progress, date, tags = [] } = project;
        return {
          id: uniqueId(),
          title,
          person,
          status: health,
          progress,
          time: moment(date, "YYYY-MM-DD").format(),
          labels: findLabelIds(labels, tags),
        };
      }),
      labels,
    }));
}

export function validateProject(project: any): Result<ImportProject, string> {
  if (
    typeof project !== "object" ||
    Object.prototype.toString.call(project) !== "[object Object]"
  )
    return err("is not a JSON object");
  if (typeof project.title !== "string")
    return err("has an invalid/missing title field");
  if (typeof project.person !== "string")
    return err("has an invalid/missing person field");
  if (!statusIds.includes(project.health))
    return err(
      `has an invalid/missing health field, it must be one of: ${statusIds.join(
        ", ",
      )}`,
    );
  if (
    typeof project.date !== "string" ||
    !/^\d\d\d\d-\d\d-\d\d$/.test(project.date)
  )
    return err(
      "has an invalid/missing date field, it must be in the YYYY-MM-DD format",
    );
  if (
    typeof project.progress !== "number" ||
    project.progress < 0 ||
    project.progress > 100
  )
    return err("has an invalid/missing, it must be a number between 0 and 100");
  if (
    project.tags !== undefined &&
    (!Array.isArray(project.tags) ||
      project.tags.some((t: any) => typeof t !== "string"))
  )
    return err("has an invalid tags list, it must be an array of strings");

  return ok(project);
}

function parseFile(fileContents: string): Result<any[], string> {
  try {
    const asObj = JSON.parse(fileContents);
    return Array.isArray(asObj) ? ok(asObj) : err("isn't a JSON array");
  } catch {
    return err("is not valid JSON");
  }
}

function generateLabels(projects: ImportProject[]): Label[] {
  return projects
    .reduce((memo, { tags = [] }) => [...memo, ...tags], [])
    .map((label) => ({ title: label, id: uniqueId() }))
    .filter((x, i, self) => i === self.findIndex((y) => x.title === y.title));
}

function findLabelIds(labels: Label[], projectTags: string[]) {
  return projectTags.map(
    (tag) => (labels.find((l) => l.title === tag) as any).id,
  );
}
