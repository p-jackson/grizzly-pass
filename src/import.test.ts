vi.unmock("./unique-id");

import { parse, parseISO, format, formatISO } from "date-fns";
import { importFile, validateProject } from "./import";
import { ok } from "./result";
import type { Project, Label } from "./types";

const untaggedProjects = [
  {
    title: "Tea Swirl",
    person: "Karen Lemon",
    date: "2017-08-28",
    progress: 23,
    health: "ontrack",
  },
  {
    title: "Rake Jumper",
    person: "Lennie Apple",
    date: "2018-02-19",
    progress: 90,
    health: "onhold",
  },
  {
    title: "Empty Tag List",
    person: "Timmy no Tag",
    date: "2018-02-17",
    progress: 90,
    health: "onhold",
    tags: [],
  },
];

const taggedProjects = [
  {
    title: "Coffee Swirl",
    person: "Joe Lemon",
    date: "2017-03-15",
    progress: 13,
    health: "ontrack",
    tags: ["Apple"],
  },
  {
    title: "Rake Twister",
    person: "Alex Apple",
    date: "2017-03-15",
    progress: 50,
    health: "onhold",
    tags: ["Apple", "Orange"],
  },
];

const allProjects = [...taggedProjects, ...untaggedProjects];

function uniqueIds({
  projects,
  labels,
}: {
  projects: Project[];
  labels: Label[];
}) {
  const projectIds = projects.map(({ id }) => id);
  const labelIds = labels.map(({ id }) => id);
  const ids = [...projectIds, ...labelIds];
  return (
    ids.every((i) => typeof i === "string") && new Set(ids).size === ids.length
  );
}

// Remove ids so projects can be snapshotted (ids can be
// different each time).
function stripIds({
  title,
  projects,
  labels,
}: {
  title?: string;
  projects: Project[];
  labels: Label[];
}) {
  return {
    title,
    projects: projects.map(({ id, labels, ...rest }) => rest),
    labels: labels.map(({ id, ...rest }) => rest),
  };
}

// Force project end times to UTC time (switching the
// actual instant in time) so it can be snapshotted.
function utcTimes({
  title,
  projects,
  labels,
}: {
  title?: string;
  projects: Partial<Project>[];
  labels: Partial<Label>[];
}) {
  return {
    title,
    projects: projects.map((project) => ({
      ...project,
      time: format(project.time, "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'"),
    })),
    labels,
  };
}

describe("importFile", () => {
  it("returns invalid if the file is not an array", () => {
    expect(importFile("3").isErr()).toBe(true);
    expect(importFile("{a:1}").isErr()).toBe(true);
    expect(importFile(`"string"`).isErr()).toBe(true);
  });

  // The import format doesn't support titles
  it("always returns undefined for the title", () => {
    expect(
      importFile(JSON.stringify(allProjects)).unsafeUnwrap(),
    ).toMatchObject({
      title: undefined,
    });
  });

  it("imports projects with no labels", () => {
    const result = importFile(JSON.stringify(untaggedProjects)).unsafeUnwrap();
    expect(result).toMatchObject({
      labels: [],
      projects: [
        {
          title: "Tea Swirl",
          person: "Karen Lemon",
          time: formatISO(parse("2017-08-28", "yyyy-MM-dd", new Date())),
          progress: 23,
          status: "ontrack",
          labels: [],
        },
        {
          title: "Rake Jumper",
          person: "Lennie Apple",
          time: formatISO(parse("2018-02-19", "yyyy-MM-dd", new Date())),
          progress: 90,
          status: "onhold",
          labels: [],
        },
        {
          title: "Empty Tag List",
          person: "Timmy no Tag",
          time: formatISO(parse("2018-02-17", "yyyy-MM-dd", new Date())),
          progress: 90,
          status: "onhold",
          labels: [],
        },
      ],
    });
    expect(uniqueIds(result)).toEqual(true);
    expect(utcTimes(stripIds(result))).toMatchSnapshot();
  });

  it("imports projects with labels", () => {
    const imported = importFile(JSON.stringify(taggedProjects));

    const result = imported.unsafeUnwrap();
    expect(result).toMatchObject({
      labels: [{ title: "Apple" }, { title: "Orange" }],
      projects: [
        {
          title: "Coffee Swirl",
          person: "Joe Lemon",
          time: formatISO(parse("2017-03-15", "yyyy-MM-dd", new Date())),
          progress: 13,
          status: "ontrack",
        },
        {
          title: "Rake Twister",
          person: "Alex Apple",
          time: formatISO(parse("2017-03-15", "yyyy-MM-dd", new Date())),
          progress: 50,
          status: "onhold",
        },
      ],
    });

    expect(result.projects[0].labels.length).toBe(1);
    expect(result.projects[1].labels.length).toBe(2);

    const allLabelIds = [
      ...new Set([...result.projects[0].labels, ...result.projects[1].labels]),
    ];
    expect(allLabelIds).toEqual([result.labels[0].id, result.labels[1].id]);

    expect(uniqueIds(result)).toBe(true);
    expect(utcTimes(stripIds(result))).toMatchSnapshot();
  });

  it("interprets dates as midnight local time", () => {
    const { time } = importFile(
      JSON.stringify(allProjects.slice(0, 1)),
    ).unsafeUnwrap().projects[0];

    // Timezone offsets change depending on the date (remember daylight savings)
    // so get the local offset for the date in allProjects[0]
    const offset = parse(
      allProjects[0].date,
      "yyyy-MM-dd",
      new Date(),
    ).getTimezoneOffset();

    expect(new Date(time).getTimezoneOffset()).toBe(offset);
    expect(new Date(time).getHours()).toBe(0);
    expect(new Date(time).getMinutes()).toBe(0);
    expect(new Date(time).getSeconds()).toBe(0);
    expect(new Date(time).getMilliseconds()).toBe(0);
  });

  it("returns an array of errors for invalid projects", () => {
    const { title, ...rest } = allProjects[0];
    const err = importFile(JSON.stringify([rest]))
      .orElse((e) => ok(e))
      .unsafeUnwrap();
    expect((err as any).length).toBe(1);
  });
});

describe("validateProject", () => {
  allProjects.forEach((project, i) => {
    it(`correctly validates an importable project ${i}`, () => {
      expect(validateProject(project).unsafeUnwrap()).toEqual(project);
    });
  });

  it("rejects when project is not an object", () => {
    expect(validateProject(undefined).isErr()).toBe(true);
    expect(validateProject("str").isErr()).toBe(true);
    expect(validateProject(3).isErr()).toBe(true);
  });

  const project = untaggedProjects[0];
  Object.keys(project).forEach((key) => {
    it(`rejects project when ${key} field is missing`, () => {
      const { [key]: _, ...rest } = project;
      expect(validateProject(rest).isErr()).toBe(true);
    });
  });

  it("ignores unknown project properties", () => {
    const project = { ...allProjects[0], unknown: "prop" };
    expect(validateProject(project).unsafeUnwrap()).toEqual(project);
  });

  it("rejects a project with invalid tags field", () => {
    const project = { ...allProjects[0], tags: 3 };
    expect(validateProject(project).isErr()).toBe(true);
  });

  it("reject a project with that uses non-string tags", () => {
    const project = { ...allProjects[0], tags: ["Tag", 3, "Tag2"] };
    expect(validateProject(project).isErr()).toBe(true);
  });

  it("rejects invalid project `health` properties", () => {
    const project = { ...allProjects[0], health: "wrong" };
    expect(validateProject(project).isErr()).toBe(true);
  });
});
