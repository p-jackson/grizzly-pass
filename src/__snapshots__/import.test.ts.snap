// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`importFile > imports projects with labels 1`] = `
{
  "labels": [
    {
      "title": "Apple",
    },
    {
      "title": "Orange",
    },
  ],
  "projects": [
    {
      "person": "Joe Lemon",
      "progress": 13,
      "status": "ontrack",
      "time": "2017-03-15T00:00:00.000Z",
      "title": "Coffee Swirl",
    },
    {
      "person": "Alex Apple",
      "progress": 50,
      "status": "onhold",
      "time": "2017-03-15T00:00:00.000Z",
      "title": "Rake Twister",
    },
  ],
  "title": undefined,
}
`;

exports[`importFile > imports projects with no labels 1`] = `
{
  "labels": [],
  "projects": [
    {
      "person": "Karen Lemon",
      "progress": 23,
      "status": "ontrack",
      "time": "2017-08-28T00:00:00.000Z",
      "title": "Tea Swirl",
    },
    {
      "person": "Lennie Apple",
      "progress": 90,
      "status": "onhold",
      "time": "2018-02-19T00:00:00.000Z",
      "title": "Rake Jumper",
    },
    {
      "person": "Timmy no Tag",
      "progress": 90,
      "status": "onhold",
      "time": "2018-02-17T00:00:00.000Z",
      "title": "Empty Tag List",
    },
  ],
  "title": undefined,
}
`;
